test/e2e/instrumentation/monitoring/prometheus.go

```diff
@@ -0,0 +1,385 @@
+		defer consumer.CleanUp()
+		retryUntil(func() bool {
+			return validateQueryValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, testTimeout)
+	})
+	It("should scrape metrics from annotated services.", func() {
+		var customMetricTargetValue float64 = 100
+		var consumerName string = "prometheus-custom-service-metric"
+		var requiredSampleCount int = 1
+		var annotatePods, annotateService bool = false, true
+		query := prometheusServiceCustomMetricQuery(f.Namespace.Name, consumerName)
+		consumer := exportCustomMetric(f, consumerName, int(customMetricTargetValue), annotatePods, annotateService)
+		defer consumer.CleanUp()
+		retryUntil(func() bool {
+			return validateQueryValues(f.ClientSet, query, customMetricTargetValue, requiredSampleCount, prometheusMetricErrorTolerance)
+		}, testTimeout)
+	})
+})
+
+func consumeCPUResources(f *framework.Framework, consumerName string, cpuUsage int) *common.ResourceConsumer {
+	return common.NewDynamicResourceConsumer(consumerName, f.Namespace.Name, common.KindDeployment, 1, cpuUsage,
+		memoryUsed, 0, int64(cpuUsage), memoryLimit, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func exportCustomMetric(f *framework.Framework, consumerName string, metricValue int, annotatePods, annotateService bool) *common.ResourceConsumer {
+	var podAnnotations, serviceAnnotations map[string]string
+	if annotatePods {
+		podAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	if annotateService {
+		serviceAnnotations = map[string]string{
+			"prometheus.io/scrape": "true",
+			"prometheus.io/path":   "/Metrics",
+			"prometheus.io/port":   "8080",
+		}
+	}
+	return common.NewMetricExporter(consumerName, f.Namespace.Name, podAnnotations, serviceAnnotations, metricValue, f.ClientSet, f.InternalClientset, f.ScalesGetter)
+}
+
+func prometheusCPUQuery(namespace, podNamePrefix string, rate time.Duration) string {
+	return fmt.Sprintf(`sum(irate(container_cpu_usage_seconds_total{namespace="%v",pod_name=~"%v.*"}[%vm]))`,
+		namespace, podNamePrefix, int64(rate.Minutes()))
+}
+
+func prometheusServiceCustomMetricQuery(namespace, service string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%v",kubernetes_name="%v"})`, namespace, service)
+}
+
+func prometheusPodCustomMetricQuery(namespace, podNamePrefix string) string {
+	return fmt.Sprintf(`sum(QPS{kubernetes_namespace="%s",kubernetes_pod_name=~"%s.*"})`, namespace, podNamePrefix)
+}
+
+func retryUntil(predicate func() bool, timeout time.Duration) {
+	startTime := time.Now()
+	for {
+		if predicate() {
+			return
+		}
+		if time.Since(startTime) >= testTimeout {
+			break
+		}
+		time.Sleep(prometheusSleepBetweenAttempts)
+	}
+	framework.Failf("monitoring using prometheus test failed")
```

Comment:
Can you make the error message more useful, e.g. number of nodes with metrics vs all nodes, actual metric value vs expected, etc?